### DISCUSSION TITLE
Using tikinter from IronPython

### DISCUSSION THREAD

**Discussion by davecalkinshexagon** [IronLanguages/ironpython3#1742] [Q&A] [answered]
We're using IronPython v3.4.0 from a C# .NET Framework 4.8 desktop application running on Windows 10 64-bit.

A user of our application has requested the ability to create some UI elements using tkinter.  It appears tikinter is not included with IronPython, however, it is provided with a regular python install.

I have Python 3.11.4 installed on my machine.  I successfully ran the below python script using the installed python directly (to verify the correctness of the script).  This opens a small empty window.

```
import tkinter as tk
root = tk.Tk()
root.mainloop()
```

If I try and run the above script from inside our app using IronPython, I get an error, "No module named 'tkinter'", as expected.

When we setup the IronPython engine (PythonExecutionService.Engine) in our app, we call SetSearchPaths() and add the Libs directory provided with IronPython (which is a sub-directory in our app install directory).  We name it "PythonLibs" but it's from the IronPython distro.  We do this using the below code.

```
var paths = Engine.GetSearchPaths();
var location = Assembly.GetAssembly(typeof(PythonExecutionService)).Location;
var dir = Path.GetDirectoryName(location) + "\\PythonLibs";
paths.Add(dir);
Engine.SetSearchPaths(paths);
```

My first attempt to get this working was to try and add the Libs directory from the regular Python installation.  I did this after we add the IronPython Libs directory so it will look first in what we get with IronPython and only after that go to the regular Python install.  For now I hard-coded the install location just to see if I could get it to work

```
paths.Add(dir); // from prev. code snippet
paths.Add(@"D:\python\Python311\Lib"); // added this to also search the regular python install
Engine.SetSearchPaths(paths);
```

After this change, I tried again to run the script from our app.  This time, I get a slightly different error, "No module named '_tkinter'".  Notice the leading underscore in the name.  It looks like this is a native library (.lib) which is in the regular python install directory but in a libs (plural) sub-directory.

I tried also adding the libs sub-directory but get the same error.

```
paths.Add(dir); // from prev. code snippet
paths.Add(@"D:\python\Python311\Lib"); // added this to also search the regular python install
paths.Add(@"D:\python\Python311\libs"); // also added this for the _tkinter.lib
Engine.SetSearchPaths(paths);
```

Finally, I tried adding the libs directory to the PATH environment variable before calling CreateEngine() with the below.

```
var path = Environment.GetEnvironmentVariable("PATH", EnvironmentVariableTarget.Process);
path += @";D:\python\Python311\libs";
Environment.SetEnvironmentVariable("PATH", path, EnvironmentVariableTarget.Process);
```

I still get the same error.  It's interesting that the error changed from 'tkinter' to '_tkinter' after adding the python Lib directory.  This suggests to me that it is finding tkinter but that the secondary dependency (_tkinter) is not found.

Any suggestions on how to get this to work?  Has anyone else had luck using tkinter from IronPython?

**Comment by slozier** [comment, marked as answer]
`_tkinter` is a C extension module and unfortunately IronPython can't handle those.

**Comment by davecalkinshexagon** [comment]
@slozier thanks for the reply!  Any suggestions for a GUI that would work with IronPython?  Our users would like to be able to write Python scripts which run in our app and which can prompt the user for inputs, etc.

**Reply by Muny** [reply]
FWIW, in my script hosting app I reimplimented the tkinter.simpledialog class with WPF windows. The API is identical and the dialogs behave the same. This was enough to satisfy the needs for most of my users.

Some of the more complex use cases use WPF directly, which is surprisingly usable from IronPython scripts.

Neither option is ideal, but they do work alright. It would be really neat if IronPython worked with Python C extension modules.

**Reply by davecalkinshexagon** [reply]
@Muny thanks for the suggestion!  Was thinking using WPF may be the way to go.  Our app uses WPF and was thinking the IronPython script should have access to it.  Tried the below.

```
import clr
clr.AddReference("PresentationFramework")

from System.Windows import Application

app = Application()
```

It reports that it **cannot import name Application**.  I'm sure I'm missing something obvious.  Any bread crumbs to get a really basic hello world WPF app working with IronPython?
